go/vt/wrangler/keyspace.go

```diff
@@ -337,138 +383,88 @@ func (wr *Wrangler) migrateServedTypesLocked(ctx context.Context, keyspace strin
-				return si.UpdateDisableQueryService(ctx, topodatapb.TabletType_MASTER, nil, true)
-			}); err != nil {
-				return err
-			}
-		}
-		if err := wr.refreshMasters(ctx, sourceShards); err != nil {
-			return err
-		}
+	event.DispatchUpdate(ev, "disabling query service on all source masters")
+	if err = wr.updateShardRecords(ctx, sourceShards, nil, topodatapb.TabletType_MASTER, true); err != nil {
+		return err
+	}
+	if err := wr.refreshMasters(ctx, sourceShards); err != nil {
+		return err
+	}
 
-		event.DispatchUpdate(ev, "getting positions of source masters")
-		masterPositions, err := wr.getMastersPosition(ctx, sourceShards)
-		if err != nil {
-			return err
-		}
+	event.DispatchUpdate(ev, "getting positions of source masters")
+	masterPositions, err := wr.getMastersPosition(ctx, sourceShards)
+	if err != nil {
+		return err
+	}
 
-		event.DispatchUpdate(ev, "waiting for destination masters to catch up")
-		if err := wr.waitForFilteredReplication(ctx, masterPositions, destinationShards, filteredReplicationWaitTime); err != nil {
-			return err
-		}
+	event.DispatchUpdate(ev, "waiting for destination masters to catch up")
+	if err := wr.waitForFilteredReplication(ctx, masterPositions, destinationShards, filteredReplicationWaitTime); err != nil {
+		return err
 	}
 
-	// Check and update all source shard records.
-	// We remember if we need to refresh the state of the source tablets
-	// so their query service is enabled again, for reverse migration.
-	event.DispatchUpdate(ev, "updating source shards")
-	needToRefreshSourceTablets := false
-	for i, si := range sourceShards {
-		sourceShards[i], err = wr.ts.UpdateShardFields(ctx, si.Keyspace(), si.ShardName(), func(si *topo.ShardInfo) error {
-			if err := si.UpdateServedTypesMap(servedType, cells, !reverse); err != nil {
-				return err
-			}
-			if tc := si.GetTabletControl(servedType); reverse && tc != nil && tc.DisableQueryService {
-				// this is a backward migration, where the
-				// source tablets were disabled previously, so
-				// we need to refresh them
-				if err := si.UpdateDisableQueryService(ctx, servedType, cells, false); err != nil {
-					return err
-				}
-				needToRefreshSourceTablets = true
-			}
-			if !reverse && servedType != topodatapb.TabletType_MASTER {
-				// this is a forward migration, we need to
-				// disable query service on the source shards.
-				// (this was already done for masters earlier)
-				if err := si.UpdateDisableQueryService(ctx, servedType, cells, true); err != nil {
-					return err
-				}
-			}
-			return nil
-		})
+	// Destination shards need different handling than what updateShardRecords does.
+	event.DispatchUpdate(ev, "updating destination shards")
+	for i, si := range destinationShards {
+		// Stop VReplication streams if we're migrating master (forward only).
+		ti, err := wr.ts.GetTablet(ctx, si.MasterAlias)
 		if err != nil {
 			return err
 		}
-	}
-	if needToRefreshSourceTablets {
```

Comment:
The refresh of the source shards is done at the beginning of the function as soon as their shard records are updated. This is because we have to first stop serving , then wait for replication to catch-up, and then only enable destinations. The diff tool makes this hard to see. The code reads a lot better if you just looked at it without the diff.

Reply:
Oh my comment was incorrect. It should have been: *refresh destination shards*. Before the change when migrating masters:

```
			if err := si.UpdateServedTypesMap(servedType, cells, reverse); err != nil {
				return err
			}
			if tc := si.GetTabletControl(servedType); !reverse && tc != nil && tc.DisableQueryService {
				// This is a forwards migration, and the
				// destination query service was already in a
				// disabled state. We need to enable and force
				// a refresh, otherwise it's possible that both
				// the source and destination will have query
				// service disabled at the same time, and
				// queries would have nowhere to go.
				if err := si.UpdateDisableQueryService(ctx, servedType, cells, false); err != nil {
					return err
				}
				needToRefreshDestinationTablets = true
			}
        // .... then
	if needToRefreshDestinationTablets {
		event.DispatchUpdate(ev, "refreshing destination shard tablets so they restart their query service")
		for _, si := range destinationShards {
			wr.RefreshTabletsByShard(ctx, si, []topodatapb.TabletType{servedType}, cells)
		}
	}

```

After the change, I don't see this happening. I'm only seeing refresh destination master (which existent version also does).

Reply:
We synced on this offline, and got to the conclusion that the slight new behavior makes more sense than the original logic.